### PR TITLE
Changes to improve performance of min/max.

### PR DESCRIPTION
**Pull request by shiksha11**
Fixes issue #16249.

```javascript
>>>Min(*[1,1,3,67,45,45,37,39,34,57,38,37,39,23,4,523,23,56,23,45,56,23,4,56,90,56,34,23,45,67,78,34,23,45,67,34,56,78,23,45,56,78,92,34,54,22,33,44,55,11,3,4,7,2,6,9,45,34,23,23],evaluate = False)
```
Computation time for above input:
Before Changes: --- 0.018165111541748047 seconds ---
After Changes: --- 0.009451150894165039 seconds ---

* functions
    * Made changes  in _find_localzeroes function to improve computation time when the number of arguments in Min/Max is large.It improves computation time when the type of arguments passed is  not <class 'sympy.core.symbol.Symbol'>.

### CODE REVIEW

File: sympy/functions/elementary/miscellaneous.py [hunk 1]
```diff
@@ -338,25 +338,22 @@ def real_root(arg, n=None, evaluate=None):
 
 class MinMaxBase(Expr, LatticeOp):
     def __new__(cls, *args, **assumptions):
-        evaluate = assumptions.pop('evaluate', True)
```
Comment (oscargus) [inline sympy/functions/elementary/miscellaneous.py]: This will still be needed to cover the case the `evaluate` is not passed.

**Comment by shiksha11** [comment]
can you please suggest me how should i proceed to fix the cases which are
failing now?

**Comment by shiksha11** [comment]
Can you suggest me a case when the generator function generates no value in
_find_localzeros.  Because i am getting list index out of range error.

py3.7.egg/sympy/functions/elementary/miscellaneous.py", line 541, in
_find_localzeros
        if str(type(generator_values[0])) == "<class
'sympy.core.symbol.Symbol'>":
    IndexError: list index out of range

**Comment by shiksha11** [comment]
py3.7.egg/sympy/functions/elementary/miscellaneous.py", line 541, in _find_localzeros
        if str(type(generator_values[0])) == "<class 'sympy.core.symbol.Symbol'>":
    IndexError: list index out of range

**Comment by shiksha11** [comment]
> py3.7.egg/sympy/functions/elementary/miscellaneous.py", line 541, in _find_localzeros
> if str(type(generator_values[0])) == "<class 'sympy.core.symbol.Symbol'>":
> IndexError: list index out of range

I am getting this error. can anyone suggest me a case when the generator will not generate any value?

**Comment by oscargus** [comment]
Here are some cases that cause problems, so maybe you can figure something out from that:
```
Min(x, 2, x**3)
 Union(S.Complexes, FiniteSet(oo))
```
where the latter gives the error you are talking about.

It may be related to https://github.com/sympy/sympy/pull/15809 so Min/Max without arguments, but I am not sure if it is related.

**Comment by shiksha11** [comment]
> Here are some cases that cause problems, so maybe you can figure something out from that:
> 
> ```
> Min(x, 2, x**3)
>  Union(S.Complexes, FiniteSet(oo))
> ```
> where the latter gives the error you are talking about.
> 
> It may be related to #15809 so Min/Max without arguments, but I am not sure if it is related.

I made changes to fix these.

**Comment by oscargus** [comment]
Probably you would like to use `.is_symbol`, `.is_number` and/or `.is_Number` is the checks. Also note that not all numbers to be checked are integers.

The `is_real` check in `_rewrite_as_Piecewise` was recently removed, so I do not think we want to introduce it again.

**Comment by shiksha11** [comment]
> Probably you would like to use `.is_symbol`, `.is_number` and/or `.is_Number` is the checks. Also note that not all numbers to be checked are integers.
> 
> The `is_real` check in `_rewrite_as_Piecewise` was recently removed, so I do not think we want to introduce it again.

Yes, I can use this too. I will change that. 
Can you suggest me how can i fix the following:

 File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/functions/elementary/tests/test_miscellaneous.py", line 388, in test_rewrite_MaxMin_as_Piecewise
    assert Max(vx, vy).rewrite(Piecewise) == Piecewise((vx, vx >= vy), (vy, True))
AssertionError

**Comment by shiksha11** [comment]
Is there any difference between is_number and is_Number ?

**Comment by oscargus** [comment]
If I remember correctly `is_number` is an actual number, while `is_Number` is something that could be turned into a number, i.e. no free symbols or undefined functions, e.g. an integral that is not yet evaluated. Or the opposite.

**Comment by oscargus** [comment]
Regarding the error, I think the fix is to not change the way it is rewritten. How does it come out now? (I guess that there are some more ways to rewrite it, but the current one makes sense if nothing else.)

**Comment by jksuom** [comment]
`expr.is_Number` should mean `isinstance(expr, Number)`. I think that it was supposed to save some time.

**Comment by shiksha11** [comment]
> If I remember correctly `is_number` is an actual number, while `is_Number` is something that could be turned into a number, i.e. no free symbols or undefined functions, e.g. an integral that is not yet evaluated. Or the opposite.

Okay,thanks.

**Comment by shiksha11** [comment]
> Regarding the error, I think the fix is to not change the way it is rewritten. How does it come out now? (I guess that there are some more ways to rewrite it, but the current one makes sense if nothing else.)

By curent one do you mean the way i have written it ?

**Comment by shiksha11** [comment]
> `expr.is_Number` should mean `isinstance(expr, Number)`. I think that it was supposed to save some time.

Yes, i added is_number to identify the data type.

**Comment by shiksha11** [comment]
Can anyone review this, please ?

**Comment by jksuom** [comment]
Some tests fail, for example 

    File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/functions/elementary/tests/test_miscellaneous.py", line 388, in test_rewrite_MaxMin_as_Piecewise
    assert Max(vx, vy).rewrite(Piecewise) == Piecewise((vx, vx >= vy), (vy, True))

Those should be investigated and fixed.

**Comment by oscargus** [comment]
> The `is_real` check in `_rewrite_as_Piecewise` was recently removed, so I do not think we want to introduce it again.

I remind you about this. Is there a reason to reintroduce it?

**Comment by shiksha11** [comment]
> Some tests fail, for example
> 
> ```
> File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/functions/elementary/tests/test_miscellaneous.py", line 388, in test_rewrite_MaxMin_as_Piecewise
> assert Max(vx, vy).rewrite(Piecewise) == Piecewise((vx, vx >= vy), (vy, True))
> ```
> Those should be investigated and fixed.

okay.

**Comment by shiksha11** [comment]
> > The `is_real` check in `_rewrite_as_Piecewise` was recently removed, so I do not think we want to introduce it again.
> 
> I remind you about this. Is there a reason to reintroduce it?

I just I want to do a check to test that the given value is numeric or not. Can you suggest me any way I can do that other than is_number ?

**Comment by oscargus** [comment]
I am talking about using `is_real` which is a completely different issue.

Using `is_number` is fine.

**Comment by shiksha11** [comment]
I am not able to give time because my exams are going on. I will definitely continue fixing this after my exams are over(by 2nd May).

**Comment by czgdp1807** [comment]
On master, with `python3`
```python
from sympy import *
import time
start = time.time()
s = Min(*[1,1,3,67,45,45,37,39,34,57,38,37,39,23,4,523,23,56,23,45,56,23,4,56,90,56,34,23,45,67,78,34,23,45,67,34,56,78,23,45,56,78,92,34,54,22,33,44,55,11,3,4,7,2,6,9,45,34,23,23],evaluate = False)
end = time.time()
print(end - start)
```
Result - 0.00039839744567871094s

```python
from sympy import *
import time
start = time.time()
s = Min(*symbols('x1:x50'), evaluate=False)
end = time.time()
print(end - start)
```
Result - 0.00021505355834960938s

ping @asmeurer Is it possible to close this? Please let me know what time it takes on your system.

**Comment by czgdp1807** [comment]
@asmeurer Any comments?

**Comment by czgdp1807** [comment]
Closing this as `master` is faster than changes in this PR. Thanks for your contributions.

**Comment by HAWK1704** [comment]
> Fixes issue #16249.
> 
> ```js
> >>>Min(*[1,1,3,67,45,45,37,39,34,57,38,37,39,23,4,523,23,56,23,45,56,23,4,56,90,56,34,23,45,67,78,34,23,45,67,34,56,78,23,45,56,78,92,34,54,22,33,44,55,11,3,4,7,2,6,9,45,34,23,23],evaluate = False)
> ```
> 
> Computation time for above input: Before Changes: --- 0.018165111541748047 seconds --- After Changes: --- 0.009451150894165039 seconds ---
> 
> * functions
>   
>   * Made changes  in _find_localzeroes function to improve computation time when the number of arguments in Min/Max is large.It improves computation time when the type of arguments passed is  not <class 'sympy.core.symbol.Symbol'>.

I know this PR is closed but I think I can optimize it, just in case you need some optimization on this.

**Pull request closed without merging**